### ISSUE TITLE
[kotlin] Add support for Kotlin

### ISSUE THREAD

**Issue by SUPERCILEX** [pmd/pmd#419]
Since Google announced official support for Kotlin, a lot more people are going to be using the language (including me 😀) so it would be nice to have some rules for it.

**Comment by vanniktech** [comment]
As a side note. Lint will start operating on UAST (Universal Abstract Syntax Tree) which means, that the Lint rules will also work with Kotlin. It might be possible to achieve the same here since some of the rules are working on an abstraction of the syntax, if not all (not that familiar with the internals of PMD).

**Comment by ryan-gustafson** [comment]
The method for adding new language support to PMD is well established.  The particulars are unique to the language however.  PMD is basically an AST visitation framework, with Rules being unique visitation patterns.  Some languages have custom parsers/AST created (e.g. Java), while others leverage available object graphs to wrap as AST (e.g. XML DOM).  If a canonical parser/AST is available this could be used to quickly bootstrap Kotlin language support.  Since languages evolve over time, it is useful for the AST to be relatively stable across versions of the language, or Rules may stop working.

A general outline of this process is here: See https://pmd.github.io/pmd-5.7.0/customizing/new-language.html

AFAIK, there is not currently support for meta-language Rules, so if you want a Rule that works for Java to work for Kotlin, you'd need to re-implement it for Kotlin.

**Comment by jsotuyod** [comment]
Rules ideas for the day we get this on:

https://ktlint.github.io/#rules
https://github.com/arturbosch/detekt#rulesets

**Comment by xanderblinov** [comment]
Hi guys, any news here? Are you plan to do it?

**Comment by jsotuyod** [comment]
@xanderblinov this is still on the roadmap. As previously stated, this needs either:

1. Full support for Antlr grammars (hopefully part of PMD 7 /cc @matifraga)
2. An abstraction on top of Antlr, similar to what we did for Apex

We are inclined on 1 as it's a one time cost for multiple languages we want to fully support (swift, kotlin, go...), while 2 is an effort needed per language. However, PRs are always welcomed to help get this shipped faster.

If anyone wants to contribute, reach out to me on gitter https://gitter.im/pmd/pmd

**Comment by yanex** [comment]
Kotlin developer here.
What are you willing to do with just an ANTLR grammar? IMO any sensible static analysis will be impossible without the type/reference resolution support. And trying to "emulate" (re-implement) it is really a bad idea. It's a very complex task, and the result won't reflect the reality anyway.

**Comment by xanderblinov** [comment]
@jsotuyod what do you think about directly complier call and using UAST?

**Comment by jsotuyod** [comment]
I'm hesitant to do so. It would require PMD to bundle the kotlin compiler, which is dependent on the Kotlin version in use (experimental features may break even in minor versions, and there are no compatibility promises for an eventual Kotlin 2 release, as per https://kotlinlang.org/docs/reference/compatibility.html).

Bundling several different versions means a really heavyweight PMD build, and not doing so means forcing users to stick to an outdated PMD version of upgrade their codebases to whatever Kotlin version we decide to support.

Unless IntelliJ decides to join forces with us, the same way Salesforce did for Apex / VisualForce support, and work with us to avoid these issues, I've some doubts this approach will be sustainable.

Using the Kotlin parser to get the UAST still needs mapping those nodes to our own `Node` structures for the pmd-core to be able to work on them. This is exactly what we do on Salesforce Apex code for instance, so the main difference is wether we use a grammar implementation from someone else, or their own. Both approaches have their benefits and their cons, I don't think one is clearly better over the other, so I'm ok with either on this particular aspect and it's a tradeoff I think we can work with.

On the other hand, **we do have type and reference resolution in place for JVM-based languages**. I know Kotlin can compile to other platforms, but probably most of our users for the time being are focused on Android support, so we are probably just fine were we stand at the moment.

**Comment by JLLeitschuh** [comment]
This is technically a "Fork" of the work done by [Poeschl](https://github.com/Poeschl) which is, itself, a fork of the
original project [pmd-kotlin](https://github.com/jk1/pmd-kotlin) from [jk1](https://github.com/jk1).

People are welcome to expand upon this project and make the tests pass. I had to kill a lot of the old tests because the compiler API's that were used to make the checks were removed/changed.

I'd like to publish this project eventually so I can use the CPD feature in my own personal build.

I'm also happy to roll this into the PMD project if there's an interest.

https://github.com/JLLeitschuh/pmd-kotlin

**Comment by Zhuinden** [comment]
Do I understand correctly that the latest PMD release (6.10.0) supports CPD for Kotlin? 🙏

**Comment by jsotuyod** [comment]
@Zhuinden yes, you understand perfectly.

**Comment by jsotuyod** [comment]
For everyone following this, starting from PMD 6.13.0 the CPD support for Kotlin is now complete, which means it now supports suppressions through comments and `CPD-OFF` / `CPD-ON`. [Check the docs for more detail](https://pmd.sourceforge.io/snapshot/pmd_userdocs_cpd.html#suppression).

Full Kotlin support for PMD rules will probably be part of PMD 7.0.0 (although we may have no actual rules at that point yet, but contributions are welcomed)

**Comment by matifraga** [comment]
We will also add a wiki page so you can enable full CPD support for your second favourite language ❤️ in a few steps

+ @tomidelucca @lsoncini

**Comment by Jeeppler** [comment]
@jsotuyod are there any rules for Kotlin? What one has to do to add new rules for Kotlin?

**Comment by jborgers** [comment]
Curious about the status of Kotlin support by PMD. Is it going to be in 7.0? Is there an early version to try out?

**Comment by adangel** [comment]
If someone wants to start the Kotlin for PMD approach, I would suggest to use PMD 7 as the basis, as there we have support for Antlr-based grammars. There is now a official antlr-grammar available for Kotlin - https://github.com/Kotlin/kotlin-spec
This could be used to add full support with moderate effort.

Having support for a language is one piece, the other piece would be to provide interesting and useful rules.

**Comment by jborgers** [comment]
We are looking into this. What would be a good way to contribute? Open an issue to discuss it?

**Comment by adangel** [comment]
Thanks for your offer. I think, this issue here is the issue to discuss adding support for Kotlin.
As described in my earlier comment, I'd suggest to integrate it into PMD 7 - since we have there full antlr support. Otherwise you would need to reimplement Kotlin grammar in javacc.

PMD 7 development happens in the branch [`pmd/7.0.x`](https://github.com/pmd/pmd/tree/pmd/7.0.x) - so make sure to base your PRs against this branch (and not master).

In PMD 7, there is one language that uses Antlr, where you can have a look, how this is integrated: Swift.
The source for that can be found here: https://github.com/pmd/pmd/tree/pmd/7.0.x/pmd-swift/src/main/java/net/sourceforge/pmd/lang/swift

The PMD 7 documentation page is here: https://docs.pmd-code.org/pmd-doc-7.0.0-SNAPSHOT/ - e.g. [Swift](https://docs.pmd-code.org/pmd-doc-7.0.0-SNAPSHOT/pmd_rules_swift.html) has 4 rules.

The Antlr integration documentation is not integrated yet - so the docs [Adding PMD support for a new language](https://docs.pmd-code.org/pmd-doc-7.0.0-SNAPSHOT/pmd_devdocs_major_adding_new_language.html) only describes the javacc way. But there is also a pull request #1881 which contains the new doc for Antlr-based languages.

**Comment by jborgers** [comment]
Hi Andreas,

Good to know, the PR #1881 documentation is what we need.
I cloned team-raptor/pmd, went to 7.0 branch and found I need to generate
those docs.
To generate the doc, I found need jekyll and bundle. For that I installed
ruby. For that I installed homebrew. And then I get into all kinds of
dependency errors between various gems and ruby.
Do you maybe have an easier way to get to that documentation? Thanks!

**Comment by adangel** [comment]
@jborgers I've updated the documentation and merged the PR. The doc is now available at https://docs.pmd-code.org/pmd-doc-7.0.0-SNAPSHOT/pmd_devdocs_major_adding_new_language_antlr.html

Please let us know, if there is anything wrong/unclear/doesn't work.

Also make sure to clone pmd/pmd - be careful when cloning forks, they might not be up-to-date...

**Comment by jborgers** [comment]
@adangel thanks, this is really helpful. 

How can I provide feedback on typo's/unclearness etc.?

Yes, not using a fork anymore.

**Comment by jborgers** [comment]
Hi @adangel 
Step 4.2 `./mvnw generate-source -pl pmd-swift `or `./mvnw generate-source -pl pmd-kotlin` does not work
I managed to generate code with  `../mvnw antlr4:antlr4` and got a KotlinLexer. How does that relate to the needed KotlinParser?
Can we communicate more interactively?
Thanks in advance.

**Comment by jborgers** [comment]
I see that Swift comes with a grammer Swift (Swift.g4) while Kotlin comes with a lexer grammer KotlinLexer (KotlinLexer.g4).
For swift I can generate several classes including the Parser. For Kotlin I get:
[INFO] --- antlr4-maven-plugin:4.8:antlr4 (default-cli) @ pmd-kotlin ---
[INFO] No grammars to process

**Comment by oowekyala** [comment]
> Step 4.2 ./mvnw generate-source -pl pmd-swift or ./mvnw generate-source -pl pmd-kotlin does not work


It's `generate-sources` with an s

**Comment by adangel** [comment]
> How can I provide feedback on typo's/unclearness etc.?

I've now created #3496 to collect the improvements for the doc. You can either comment here in this issue or in the PR.

> ... got a KotlinLexer. How does that relate to the needed KotlinParser?

So, PMD 6 also supports Kotlin - but only CPD (Copy Paste Detection). For CPD, only a Lexer (the first part of a parser/compiler) is required - it only tokenizes the source code.
The Parser is the next step.

So your task is to throw away the existing grammar and use the new official Kotlin grammar instead from https://github.com/Kotlin/kotlin-spec ...

Make sure to set the property `<antlr4.visitor>true</antlr4.visitor>` in the pom.xml file, as otherwise a parser probably won't be generated.

> Can we communicate more interactively?

You can try to use https://gitter.im/pmd/pmd - but don't expect someone to be around all the time. We do this in our free spare time.

**Comment by jborgers** [comment]
Okay, good. I will report typo's etc over there and if I have a question here.

Okay, good to know.

Great, I can join the room.

**Comment by jborgers** [comment]
I got a bit further, I included KotlinLexer.g4 and KotlinParser.g4 from kotlin-spec. I can generate code, however the build fails.

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-antrun-plugin:1.8:run (antlr-cleanup) on project pmd-kotlin: An Ant BuildException has occured: The following error occurred while executing this line:
[ERROR] /Users/jeroen/IdeaProjects/pmd-pmd/antlr4-wrapper.xml:80: Replace: source file /Users/jeroen/IdeaProjects/pmd-pmd/pmd-kotlin/target/generated-sources/antlr4/net/sourceforge/pmd/lang/kotlin/ast/KotlinVisitor.java doesn't exist
[ERROR] around Ant part ...<ant antfile="/Users/jeroen/IdeaProjects/pmd-pmd/pmd-kotlin/../antlr4-wrapper.xml" target="pmd-language">... @ 4:108 in /Users/jeroen/IdeaProjects/pmd-pmd/pmd-kotlin/target/antrun/build-main.xml

The code generator generated a class with name KotlinParserVisitor.java instead, and other classes starting with KotlinParser like KotlinParserListener etc. I guess this has to do with the file name being KotlinParser.g4, different than Swift.g4, which includes both lexer and parser grammar. How to solve this?

You can replay this using my repo.
Thanks

**Comment by adangel** [comment]
So, it seems, that we need to combine the 3 g4 files from kotlin-spec into a single file called "Kotlin.g4".
According to the docs, this is possible: https://github.com/antlr/antlr4/blob/master/doc/grammars.md

> Grammars defined without a prefix on the grammar header are combined grammars that can contain both lexical and parser rules.

So, we need to make sure, our Kotlin.g4 file starts with this header:

```
grammar Kotlin;

@header {
import net.sourceforge.pmd.lang.ast.impl.antlr4.*;
import net.sourceforge.pmd.lang.ast.AstVisitor;
}

```
(The imports are important, as we'll modify later via our ant script the generated parser and missing these imports would create uncompilable code).

After that, you simply copy first KotlinParser.g4 (without its header and without the options), then KotlinLexer.g4 (without its header and without "import UnicodeClasses;") followed by UnicodeClasses.g4 (again without its header).

I'm not sure, whether you need the "*.tokens" files. As far as I have read, these are usually generated. Maybe the tokens file are used - I don't know. If you want, you can create a single file "Kotlin.tokens" - just use the content of KotlinLexer.tokens, this file seems to contain all tokens (from KotlinLexer and UnicodeClasses).
I guess, if the file "Kotlin.tokens" is used, then - when adding a new token - the old token ids stay the same, which helps in keeping the parser/tokenizer backwards compatible.

**Comment by jborgers** [comment]
Thanks! Working on it. 
I got some new errors, will work on that more.

**Comment by jborgers** [comment]
I think I also need @parser::members and options {}, translated from Swift to Kotlin.
There is however a problem I don't know how to solve: the lexer part contains 'modes' and the antlr grammers doc says:
"Only lexer grammars can contain mode specifications" 

output:
```
[INFO] --- antlr4-maven-plugin:4.8:antlr4 (antlr) @ pmd-kotlin ---
[INFO] ANTLR 4: Processing source directory /Users/jeroen/IdeaProjects/pmd-pmd/pmd-kotlin/src/main/antlr4
[INFO] Processing grammar: net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4
[ERROR] error(176): net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4:2639:24: Inside is not a recognized mode name
[ERROR] /Users/jeroen/IdeaProjects/pmd-pmd/net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4 [2639:24]: Inside is not a recognized mode name
[ERROR] error(176): net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4:2641:25: Inside is not a recognized mode name
[ERROR] /Users/jeroen/IdeaProjects/pmd-pmd/net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4 [2641:25]: Inside is not a recognized mode name
[ERROR] error(176): net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4:2936:28: LineString is not a recognized mode name
[ERROR] /Users/jeroen/IdeaProjects/pmd-pmd/net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4 [2936:28]: LineString is not a recognized mode name
[ERROR] error(176): net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4:2938:37: MultiLineString is not a recognized mode name
[ERROR] /Users/jeroen/IdeaProjects/pmd-pmd/net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4 [2938:37]: MultiLineString is not a recognized mode name
[ERROR] error(120): net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4:2940:5: lexical modes are only allowed in lexer grammars
[ERROR] /Users/jeroen/IdeaProjects/pmd-pmd/net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4 [2940:5]: lexical modes are only allowed in lexer grammars
[ERROR] error(120): net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4:2963:5: lexical modes are only allowed in lexer grammars
[ERROR] /Users/jeroen/IdeaProjects/pmd-pmd/net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4 [2963:5]: lexical modes are only allowed in lexer grammars
[ERROR] error(120): net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4:2987:5: lexical modes are only allowed in lexer grammars
[ERROR] /Users/jeroen/IdeaProjects/pmd-pmd/net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4 [2987:5]: lexical modes are only allowed in lexer grammars
[ERROR] error(120): net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4:3124:5: lexical modes are only allowed in lexer grammars
[ERROR] /Users/jeroen/IdeaProjects/pmd-pmd/net/sourceforge/pmd/lang/kotlin/ast/Kotlin.g4 [3124:5]: lexical modes are only allowed in lexer grammars
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------

```

**Comment by jborgers** [comment]
@adangel Any ideas? Help appreciated, I am stuck.

**Comment by jborgers** [comment]
I executed all steps of "Adding PMD support for a new ANTLR grammar based language" except generation of parser which doesn't succeed. Visible in my repo.

**Comment by oowekyala** [comment]
> lexical modes are only allowed in lexer grammars

What's the problem with creating a separate lexer grammar? It seems like we should just do what the error message recommends...

**Comment by jborgers** [comment]
We started with a separate lexer grammar and parser grammar, however, the code generated was a KotlinParserVisitor.java etc. (see above.) Therefore @adangel recommended to combine the two in one file, but then modes don't work.
So, maybe the code generation needs to take into account that the two are separate and omit the 'Parser' in the names of the generated classes?

**Comment by oowekyala** [comment]
> the code generated was a KotlinParserVisitor.java etc.

I assume those names are just derived from the grammar file names. Have you tried a `Kotlin.g4` for the parser and `KotlinLexer.g4` for the lexer grammar?

**Comment by jborgers** [comment]
Interesting suggestion. I just did. The parser grammar needs to be the same as the file name, so Kotlin instead of KotlinParser. The result is that indeed the KotlinVisitor etc. classes have the right name, however, the KotlinParser class is now called Kotlin. 
Renaming that to KotlinParser and running class generation again without clean, gets me further.

So maybe this renaming needs to be automated in once of the scripts or a pom?

**Comment by jborgers** [comment]
My generated KotlinParser now has errors: it calls super.acceptVisitor(visitor, data) while this method does not exist.
There is an accept(visitor) however.

**Comment by jborgers** [comment]
@adangel @oowekyala Help appreciated, I am stuck.

**Comment by oowekyala** [comment]
I'm not sure where you pushed your changes. Could you open a draft PR with the relevant branch, and give us write access (a checkbox when you open the PR)? This way we could take a look easily

**Comment by jborgers** [comment]
Okay, I just did. Let me know if easily taking a look works now

**Comment by jborgers** [comment]
Great, it seems to work now! With generated files, all compiles. 
What did you do?
Could we get the designer tool to work as well?

**Comment by oowekyala** [comment]
> What did you do? Does the documentation need to be updated, the steps?

I mostly just fixed the Ant script, I'll update the docs later.

> Could we get the designer tool to work as well?

The designer should just work, but you'll have to follow the instructions at the top of [this page](https://github.com/pmd/pmd/wiki/PMD-7.0.0) to use it with pmd 7.

**Comment by jborgers** [comment]
We could eventually managed to get designer working. We for instance also needed build-tools and IntelliJ was hanging on pmd-designer.

![image](https://user-images.githubusercontent.com/24591067/134944459-4bc65fd3-bd08-49fc-ac1a-1906617bffd2.png)

**Comment by jborgers** [comment]
But someting seems wrong in the grammar, we get a very deeply nested syntax tree for the simplest expressions:
![image](https://user-images.githubusercontent.com/24591067/134945308-f563c7f2-313a-4c84-8343-c2d7422e808f.png)
Not sure how to proceed, help appreciated.

**Comment by adangel** [comment]
The syntax tree is correct - that's how the grammar is written. @oowekyala made me aware, that ANTLR doesn't create a abstract syntax tree but instead creates a parse tree. And that is, what you are seeing. Basically everything down to the tokens is a node in the tree.

What we would need is some kind of post-processing of the parse tree, to make it look nicer and simpler to use for static analysis. E.g. the name of the function would be a perfect fit for an attribute on the node "FunctionDeclaration" (instead of having the node "T-Identifier").

The deep tree also comes from the expression production. In the Javacc grammar for Java, we already simplify this part of the tree with the help of a javacc feature (conditional tree - it creates only a tree node if there are >1 nodes).

While the tree doesn't look as good as it could, it's still usable. Most important is, that you write tests for the rules, you create (see https://pmd.github.io/latest/pmd_userdocs_extending_testing.html) because then you can ensure that the rules work in case we improve the tree later.

**Comment by stokpop** [comment]
Thanks. Yes, we will create tests for the rules, like we have for our performance Java rules.

Starting with the designer to create the first Kotlin XPath rule, we have some questions:

1. In our Java XPath rules we use the `typeIs()` function a lot, as in `pmd-java:typeIs('java.lang.String')`.  Can we use the same function in Kotlin XPath rules? Maybe `pmd-kotlin:typeIs()` is available or should be created?
2. XPath attributes are really useful for our XPath rule creation. For example, in Java rules we use `@Image` tags often. The PMD designer shows that the `@Image`  attribute and `@joinTokenText` are deprecated. `@Text` is available, but only at the lowest level, like T-Identifier. So for matching an import package: `//ImportList/ImportHeader/Identifier[starts-with(@joinTokenText, "com.netflix.hystrix")]`. The `@joinTokenText` seems very convenient. Is there an easy alternative?
3. The Kotlin code below gives Errors in the AST. Also in the ANLTR plugin in intellij there are errors starting at the first {. Could this be an issue in the Kotlin antlr definition. Maybe similar to this: https://github.com/Kotlin/kotlin-spec/issues/69. Should we raise an issue in the kotlin-spec project for this?

```kotlin
object TestFoo {
    fun testBar() = 1
}
```

**Comment by jborgers** [comment]
@adangel @oowekyala any ideas?

**Comment by oowekyala** [comment]
I'm afraid all of those are actually very real limitations, for which I have no solution...

1. the typeIs function in pmd-java uses our own implementation of a symbol and type resolver. This is a complicated component as it has to eg perform type inference. The Java type resolver in pmd 7, which was rewritten from scratch, is 40 000+ lines, excluding its tests. Building the same thing for Kotlin would be an incredibly (impossibly) massive undertaking. Kotlin has much more complicated scoping rules and type inference, it even has flow-sensitive types. I can only say that building and maintaining such a component is completely out of scope of PMD as I currently understand it. So supporting a `typeIs` function for kotlin would need, I think, maybe an integration with the kotlin compiler.
2. Attributes are also something we implement manually, in javacc languages. Antlr does not let us replace or add members to the classes it generates. The recommended way to use antlr is in fact to build (manually) another AST on top of the parse tree. In those AST classes, we could write and expose our own attributes. That would also fix the problem that the tree is too big and has irrelevant nodes, as you pointed out. However, this is not how PMD's Antlr integration is using Antlr. #1658 was based on the assumption that we could use Antlr's parse tree directly as an AST, and in retrospect I think it was pretty naive, and that continuing down this route will only cause more problems.
  One alternative would be to say that `@Image` and `@joinTokenText` are not deprecated in Antlr, because we don't have anything better anyway... But this would be setting up Antlr language modules to be second-class citizens and is honestly not a decent solution.
  Another alternative would be to manually add attributes to antlr nodes using a visitor. That would actually be pretty easy, but would only improve the situation for XPath rules. We could rely on utility classes/kotlin extensions to add API to nodes for non-xpath rules to use. But the tree would still be horribly big and undocumented.
  I'm afraid none of this sounds very compelling to me... 
3. This is probably a problem with the Antlr grammar and yes, it should be reported to kotlin-spec

**Comment by jborgers** [comment]
That was my doing, putting everything in one .g4 file like for Swift. It seemed to work.

**Comment by jborgers** [comment]
We are still working on this, see: [kotlin] Kotlin support #3505

**Comment by adangel** [comment]
Status update for this issue:
* The "basic" Kotlin support has been merged into pmd/7.0.x branch
* It will be part of PMD 7
* There are certain limitations currently that need to be improved later on
  * The AST is very deep. This is because ANTLR doesn't produce a AST, but a parse tree. We would need some post processing in order to create our own AST representation
  * Custom XPath attributes for ANTLR-based nodes are not possible
  * Type resolution is not supported for Kotlin. This would be a very complicated undertaking to reimplement all the type inference rules of Kotlin within PMD with an own type and symbol resolver. Maybe a integration with a real kotlin compiler could be an option.

I'll create issues for these points and close this one as completed.
